### ISSUE TITLE
How can i implement absolute difference in the Merge layer.

### ISSUE THREAD

**Issue by LiuAndYang** [fchollet/keras#2107]
I need to merge two layers and get the abosolte difference of them, but the merge layer only support 'sum', 'ave'.....,  these are not i need, i tried to add a 'diff' mode in Merge layer in core.py. but get wrong result. what  is the mistake in my code, just like the 'sum' mode ?  Thanks.  (ps. there have indent before s -= self.layers[i].get_output(train) , but don't show in the code, i am very new here. sorry)

  `elif self.mode == 'diff':  
         s = self.layers[0].get_output(train)
         for i in range(1, len(self.layers)):
                 s -= self.layers[i].get_output(train)
         return K.abs(s)`

**Comment by NasenSpray** [comment]
You're calculating something like `|x - y - z|`, is that what you want?

**Comment by farizrahman4u** [comment]
This is what the `LambdaMerge` layer is for.

```python
def abs_diff(X):
    s = X[0]
    for i in range(1, len(X)):
        s -= X[i]
    s = K.abs(s)
    return s

model1 = Sequential()
model1.add(Dense(input_dim=10, output_dim=10))


model2 = Sequential()
model2.add(Dense(input_dim=10, output_dim=10))

model = Sequential()
model.add(LambdaMerge([model1, model2], abs_diff))
model.add(Dense(10))

model.compile(loss='mse', optimizer='sgd')
```

**Comment by LiuAndYang** [comment]
@NasenSpray  yes, just like `|X - Y|`, but I have only two element to compute their absolute difference.

**Comment by LiuAndYang** [comment]
@farizrahman4u That works well right now ! Thank you so much~

**Comment by LiuAndYang** [comment]
@farizrahman4u  After several iterations， the loss becomes `nan`, What may lead to that?

**Comment by NasenSpray** [comment]
@LiuAndYang `∂/∂x |x-y| = (x-y)/|x-y|` and that's NaN when `x = y`

**Comment by LiuAndYang** [comment]
@NasenSpray  Thanks! I see, I have seen the absolute difference appeared in many papers. Can you give me some advise about how did they handle this problem ?  Add a small number in denominator would OK ? I am afraid that operation influence the final results, Thank you again~

**Comment by NasenSpray** [comment]
How about root square difference, i.e. `K.sqrt(K.square(x-y) + 1e-7)`, or just square difference?

**Comment by LiuAndYang** [comment]
@NasenSpray  Thanks, I will try this method. Only add a small number seems lead to worse results.

**Comment by Imorton-zd** [comment]
Hi, @LiuAndYang @NasenSpray @farizrahman4u,  when I only want to get the minimum loss of `sqrt(square(x-y) + 1e-7)`, I have no `label` in my codes as following in the method `fit`. How can I do? Thanks.
```
def sqrt_diff(X):
    s = X[0]
    for i in range(1, len(X)):
        s -= X[i]
    s = K.sqrt(K.square(s) + 1e-7)
    return s
    
def train_merge_model(train_q,train_a):
         
    print('Build model...')
    model0 = Sequential()
    model0.add(LSTM(lstm_output_dims,input_shape=(maxlen,embedding_dims)))
    model0.add(Dropout(0.5))
    
    model1 = Sequential()
    model1.add(LSTM(lstm_output_dims,input_shape=(maxlen,embedding_dims)))
    model1.add(Dropout(0.5))
    
    modelmerge = Sequential()
    modelmerge.add(LambdaMerge([model0, model1], abs_diff))
    modelmerge.add(Dropout(0.5))
  
    modelmerge.compile(loss='mse', optimizer='rmsprop')

    hist = modelmerge.fit([train_q,train_a], label, batch_size=batch_size, nb_epoch=10, verbose=1, show_accuracy=True)
    print (hist.history)
```

**Comment by Imorton-zd** [comment]
@simonhughes22 @EderSantana @wxs @gautamb85 Would you help me solve the problem above? Any opinions would be appreciated!

**Comment by EderSantana** [comment]
keras still does not support objective functions without desired. So you'll have to hack your way around it. Assume that you have the output of lambda merge that gives your the abs-diff, you could write a cost function like so:
```python
def abs_diff_cost(y_true, y_pred):
    # y_true is just a place holder
    cost = K.mean(y_pred, axis=-1) + 0. * y_true.sum()  # this is for theano to think y_true is doing something
    return cost
```

when you call `model.fit(X, Y)` make sure `Y` has the same shape as what you expect the outputs of your `abs_diff` layer to be because Keras will check for size compatibility

**Comment by Imorton-zd** [comment]
@EderSantana Thanks for your reply! I may not understand your meaning absolutely. In fact, I am going to do unsupervised learning with keras. Now, I have two types of datasets, which contains similar semantics. So I want to first encode (such as lstm) two documents from two datasets separately, then get the absolute difference of them. As long as the loss of the absolute difference achieve minimum, the representations of two documents from encoder (such as lstm) are my want. In this process, there are no need of labels of samples. Can't keras implement this idea?

**Comment by EderSantana** [comment]
hmm... this seems the job for a `regularization` check this to see how to add a regularization to your cost function: https://github.com/fchollet/keras/blob/master/keras/layers/core.py#L1177-L1210
See that you simply have to define a layer and add an element to `self.regularizers`. You may have to write your own layer to do that maybe.

But yeah, this is how I think I would do that with Keras: create a new simple layer that has only one job: add one cost to `self.regularizers`.
